src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -514,17 +515,28 @@ case class FilterVariants(
   }
 }
 
-case class KeyTableType(rowType: TStruct, key: Array[String]) extends BaseType
+case class KeyTableValue(typ: KeyTableType, localValue: KTLocalValue, rvd: RVD) {
+  def rdd: RDD[Row] = {
+    val localRowType = typ.rowType
+    rvd.rdd.map { rv => new UnsafeRow(localRowType, rv.region.copy(), rv.offset) }
+  }
+}
+
+case class KeyTableType(rowType: TStruct, key: Array[String], globalType: TStruct) extends BaseType
+
+object KeyTableIR {
+  def optimize(ir: KeyTableIR): KeyTableIR = ir
+}
 
 abstract sealed class KeyTableIR extends BaseIR {
   def typ: KeyTableType
 
-  def execute(hc: HailContext): RDD[RegionValue]
+  def execute(hc: HailContext): KeyTableValue
 }
 
 case class KeyTableLiteral(
   typ: KeyTableType,
```

Comment:
This shouldn't take a type, because `value` carries one.  You just need to add `def typ: KeyTableType = value.typ`.